server/app/actions/ProgramFastForwardApplicationAction.java

```diff
@@ -0,0 +1,64 @@
+package actions;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.ProfileUtils;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import play.mvc.Action;
+import play.mvc.Http.Request;
+import play.mvc.Result;
+import play.routing.Router;
+import repository.ApplicationRepository;
+import repository.ProgramRepository;
+
+/**
+ * Check if the current application is associated with the latest version of a program. If it is
+ * continue on to the original page as normal. If the application is on an older version update the
+ * application to the latest version and redirect to the application review page.
+ */
+public class ProgramFastForwardApplicationAction extends Action.Simple {
+  private final ProfileUtils profileUtils;
+  private final ProgramRepository programRepository;
+  private final ApplicationRepository applicationRepository;
+
+  @Inject
+  public ProgramFastForwardApplicationAction(
+      ProfileUtils profileUtils,
+      ProgramRepository programRepository,
+      ApplicationRepository applicationRepository) {
+    this.profileUtils = checkNotNull(profileUtils);
+    this.programRepository = checkNotNull(programRepository);
+    this.applicationRepository = checkNotNull(applicationRepository);
+  }
+
+  @Override
+  public CompletionStage<Result> call(Request request) {
+    String routePattern = request.attrs().get(Router.Attrs.HANDLER_DEF).path();
+    RouteExtractor routeExtractor = new RouteExtractor(routePattern, request.path());
+
+    long programId = routeExtractor.getParamLongValue("programId");
+    long applicantId = getApplicantId(request, routeExtractor);
+
+    long latestProgramId = programRepository.getMostRecentActiveProgramId(programId);
+
+    if (latestProgramId > programId) {
+      applicationRepository.updateDraftApplicationProgram(applicantId, latestProgramId);
```

Comment:
I am aware this does not do the `checkApplicantAuthorization` or `checkProgramAuthorization` checks like in the controllers.

That will be addressed in a separate PR and in a separate reusable action.